opm/simulators/wells/MultisegmentWellGeneric.cpp

```diff
@@ -480,130 +480,38 @@ computeBhpAtThpLimitProd(const std::function<std::vector<double>(const double)>&
-        double f_low = fflo(low);
-        double f_high = fflo(high);
-        deferred_logger.debug("computeBhpAtThpLimitProd(): well = " + baseif_.name() +
-                              "  low = " + std::to_string(low) +
-                              "  high = " + std::to_string(high) +
-                              "  f(low) = " + std::to_string(f_low) +
-                              "  f(high) = " + std::to_string(f_high));
-        int adjustments = 0;
-        const int max_adjustments = 10;
-        const double adjust_amount = 5.0 * unit::barsa;
-        while (f_low * f_high > 0.0 && adjustments < max_adjustments) {
-            // Same sign, adjust high to see if we can flip it.
-            high += adjust_amount;
-            f_high = fflo(high);
-            ++adjustments;
-        }
-        if (f_low * f_high > 0.0) {
-            if (f_low > 0.0) {
-                // Even at the BHP limit, we are injecting.
-                // There will be no solution here, return an
-                // empty optional.
-                deferred_logger.warning("FAILED_ROBUST_BHP_THP_SOLVE_INOPERABLE",
-                                        "Robust bhp(thp) solve failed due to inoperability for well " + baseif_.name());
-                return std::optional<double>();
-            } else {
-                // Still producing, even at high bhp.
-                assert(f_high < 0.0);
-                bhp_max = high;
-            }
-        } else {
-            // Bisect to find a bhp point where we produce, but
-            // not a large amount ('eps' below).
-            const double eps = 0.1 * std::fabs(table.getFloAxis().front());
-            const int maxit = 50;
-            int it = 0;
-            while (std::fabs(f_low) > eps && it < maxit) {
-                const double curr = 0.5*(low + high);
-                const double f_curr = fflo(curr);
-                if (f_curr * f_low > 0.0) {
-                    low = curr;
-                    f_low = f_curr;
-                } else {
-                    high = curr;
-                    f_high = f_curr;
-                }
-                ++it;
-            }
-            bhp_max = low;
-        }
-        deferred_logger.debug("computeBhpAtThpLimitProd(): well = " + baseif_.name() +
-                              "  low = " + std::to_string(low) +
-                              "  high = " + std::to_string(high) +
-                              "  f(low) = " + std::to_string(f_low) +
-                              "  f(high) = " + std::to_string(f_high) +
-                              "  bhp_max = " + std::to_string(bhp_max));
-    }
+    auto fflo = [&flo, &frates](double bhp) { return flo(frates(bhp)); };
+    auto bhp_max = this->bhpMax(fflo, controls.bhp_limit, maxPerfPress, table.getFloAxis().front(), deferred_logger);
 
+    // could not solve for the bhp-point, we could not contrinue to find the bhp
+    if (!bhp_max.has_value()) {
+        return {};
```

Comment:
Further down, std::nullopt is used, I prefer that. Also, add a debug message.

Reply:
Yes, `std::nullopt` is probably safer.